### PR TITLE
Require CMake >=3.16

### PR DESCRIPTION
**Pull request by masterleinad**
In accordance with https://github.com/trilinos/Trilinos/issues/8401, we also decided to require CMake 3.16 or higher.

### CODE REVIEW

File: cmake/Modules/CudaToolkit.cmake [hunk 1]
```diff
@@ -481,75 +481,17 @@ if(CMAKE_CUDA_COMPILER_LOADED AND NOT CUDAToolkit_BIN_DIR AND CMAKE_CUDA_COMPILE
...
+  target_link_libraries(${ARGN})
+endfunction()
+
+function(import_target_link_directories)
+  target_link_directories(${ARGN})
+endfunction()
+
+function(import_target_include_directories)
+  target_include_directories(${ARGN})
+endfunction()
```
Comment (dalg24) [inline cmake/Modules/CudaToolkit.cmake:494]: This means there is not point on having these functions any more

File: cmake/kokkos_functions.cmake [hunk 2]
```diff
@@ -230,9 +228,7 @@ MACRO(kokkos_import_tpl NAME)
   # I have still been getting errors about ROOT variables being ignored
   # I'm not sure if this is a scope issue - but make sure
   # the policy is set before we do any find_package calls
-  IF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12.0")
-    CMAKE_POLICY(SET CMP0074 NEW)
-  ENDIF()
+  CMAKE_POLICY(SET CMP0074 NEW)
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake:231]: I actually wonder if we still need that guy

File: CMakeLists.txt [hunk 3]
```diff
@@ -72,7 +72,7 @@ ENDFUNCTION()
 LIST(APPEND CMAKE_MODULE_PATH cmake/Modules)
 
 IF(NOT KOKKOS_HAS_TRILINOS)
-  cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
+  cmake_minimum_required(VERSION 3.17.1 FATAL_ERROR)
```
Comment (dalg24) [inline CMakeLists.txt:75]:
```suggestion
  cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
```
unless someone can explain what is wrong with 3.17.0

File: .jenkins [hunk 4]
```diff
@@ -285,7 +285,7 @@ pipeline {
                         dockerfile {
                             filename 'Dockerfile.nvcc'
                             dir 'scripts/docker'
-                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:10.1-devel --build-arg CMAKE_VERSION=3.15.5'
+                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:10.1-devel --build-arg CMAKE_VERSION=3.17.3'
```
Comment (dalg24) [inline .jenkins:288]: Drop the argument
```suggestion
                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:10.1-devel'
```

File: scripts/docker/Dockerfile.clang [hunk 5]
```diff
@@ -9,7 +9,7 @@ RUN apt-get update && apt-get install -y \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
 
-ARG CMAKE_VERSION=3.10.3
+ARG CMAKE_VERSION=3.17.3
```
Comment (dalg24) [inline scripts/docker/Dockerfile.clang:12]: Use `3.16.8` (or `3.16.0`) in all the dockerfiles you edited

File: .travis.yml [hunk 6]
```diff
@@ -45,6 +45,26 @@ matrix:
       compiler: gcc
       env: CMAKE_BUILD_TYPE=Release BACKEND="OPENMP"
 
+# Install newer CMake. The distribution comes with CMake 3.12.4 but we require at least 3.17.1
```
Comment (dalg24) [inline .travis.yml:48]: Update

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — APPROVED** [review]

File: cmake/kokkos_functions.cmake (hunk 2 above)
Comment (masterleinad) [reply cmake/kokkos_functions.cmake:231]: https://cmake.org/cmake/help/latest/policy/CMP0074.html still says
> [...] 
This policy was introduced in CMake version 3.12. CMake version 3.19.2 warns when the policy is not set and uses OLD behavior. Use the cmake_policy() command to set it to OLD or NEW explicitly.

Comment (dalg24) [reply cmake/kokkos_functions.cmake:231]: Look at the comment above that block of code.  The policy is being set in the main CMakeLists.txt of the package

**Review by crtrott — APPROVED** [review]

File: .travis.yml [hunk 7]
```diff
@@ -45,6 +45,26 @@ matrix:
       compiler: gcc
       env: CMAKE_BUILD_TYPE=Release BACKEND="OPENMP"
 
+# Install newer CMake. The distribution comes with CMake 3.12.4 but we require at least 3.16
+install:
+  - CMAKE_VERSION=3.17.1
```
Comment (jrmadsen) [inline .travis.yml:50]: Why not:

```
DISTRIB_CODENAME=$(cat /etc/lsb-release | grep DISTRIB_CODENAME | awk -F '=' '{print $NF}')
wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc | apt-key add -
apt-add-repository "deb https://apt.kitware.com/ubuntu/ ${DISTRIB_CODENAME} main"
```

**Review by DavidPoliakoff — APPROVED** [review]
Good *and* net negative lines

File: .travis.yml (hunk 7 above)
Comment (masterleinad) [reply .travis.yml:50]: I just copied from our `Dockerfiles`. We could try your suggestions but it doesn't matter as long as we don't activate `Travis` again.

**Comment by masterleinad** [comment]
Lots of approvals including @dalg24 and @crtrott. Awaiting merging. If requested I can squash before.

**Pull request merged**